File: internal/server/assets/flow.js
// A warehouse build drawn as what it is.
//
// The landscape next door draws code: a path that descends into a call and
// comes back, where closure is the shape you read it by. SQL has no returns.
// So this draws a DAG instead — layered left to right in build order, every
// table strictly to the right of everything it reads, and every arrow labelled
// with how the rows were matched. The two facts a call arrow cannot carry are
// the join type and the row count, and they are the two a warehouse reader
// asks for first.

function drawFlow() {
  const flow = DATA.flow;
  const svg = document.getElementById('svg');
  svg.innerHTML = '';
  document.getElementById('closed').textContent = '';

  const nodes = relayerOutsideTables(flow.nodes || [], flow.links || []);
  const links = flow.links || [];
  if (!nodes.length) {
    svg.setAttribute('height', 40);
    svg.appendChild(el('text', { x: 8, y: 24, class: 'blabel' }, 'Nothing in the DAG. Run: plum ingest'));
    return;
  }

  // Layout is measured, not assumed. Every box is as tall as it has facts to
  // state, so a fixed row height either crushes the tall ones together or
  // wastes a screen on the short ones. Columns are packed by real heights and
  // centred against each other, which also keeps the arrows short.
  const BOXW = 232, GAPX = 168, GAPY = 40, PADX = 28, PADY = 52;
  const COL = BOXW + GAPX;
  const byLayer = new Map();
  for (const n of nodes) {
    if (!byLayer.has(n.layer)) byLayer.set(n.layer, []);
    byLayer.get(n.layer).push(n);
  }
  const layers = [...byLayer.keys()].sort((a, b) => a - b);
  orderColumns(layers, byLayer, links);

  const stack = layers.map(L => {
    const col = byLayer.get(L);
    let h = 0;
    const heights = col.map(n => boxHeight(n));
    heights.forEach(v => { h += v + GAPY; });
    return { col, heights, total: h - GAPY };
  });
  const tallest = Math.max(...stack.map(s => s.total));

  const pos = new Map();
  stack.forEach((s, li) => {
    let y = PADY + (tallest - s.total) / 2;   // centre each column against the tallest
    s.col.forEach((n, i) => {
      pos.set(n.symbol, { x: PADX + li * COL, y, h: s.heights[i], node: n });
      y += s.heights[i] + GAPY;
    });
  });
  const width = PADX * 2 + (layers.length - 1) * COL + BOXW;
  const height = PADY * 2 + tallest;
  svg.setAttribute('width', width);
  svg.setAttribute('height', height);
  svg.setAttribute('viewBox', `0 0 ${width} ${height}`);

  // Arrows first, so the tables sit on top of them.
  // Several arrows can land on one table. Their labels are spread along the
  // span rather than all sitting at the midpoint, where they would overlap into
  // an unreadable pile.
  const arriving = new Map();
  for (const l of links) arriving.set(l.to, (arriving.get(l.to) || 0) + 1);
  const placed = new Map();
  const labels = [];

  for (const l of links) {
    const a = pos.get(l.from), b = pos.get(l.to);
    if (!a || !b) continue;
    const x1 = a.x + BOXW, y1 = a.y + a.h / 2;
    const x2 = b.x, y2 = b.y + b.h / 2;
    const k = placed.get(l.to) || 0;
    placed.set(l.to, k + 1);
    const spread = arriving.get(l.to) > 1 ? (k - (arriving.get(l.to) - 1) / 2) * 24 : 0;
    const mid = (x1 + x2) / 2 + spread;
    const d = `M${x1},${y1} C${mid},${y1} ${mid},${y2} ${x2},${y2}`;
    // A dependency dbt does not know about is drawn differently, because the
    // whole hazard is that it looks like part of the DAG and is not.
    const cls = !l.in_dag ? 'link-outside' : (l.relation === 'from' ? 'link-from' : 'link-join');
    svg.appendChild(el('path', { d, class: 'flowlink ' + cls, fill: 'none', 'marker-end': 'url(#arrow)' }));

    // Labels sit in the gutter immediately left of the table they arrive at,
    // not at the midpoint of the span. An arrow that crosses two layers has its
    // midpoint inside another column, and a label there lands on top of an
    // unrelated table.
    const label = linkLabel(l);
    if (label) {
      labels.push({
        x: x2 - 16, // clear of the arrowhead
        y: y2 + (k - (arriving.get(l.to) - 1) / 2) * 36 - 4,
        label, rows: l.rows,
      });
    }
  }
  // Every label goes on after every arrow. Drawing them per link meant a later
  // link's arrow painted over an earlier link's text — which read as a typo in
  // the join key rather than as an overlap.
  for (const t of labels) {
    const rows = t.rows ? fmtRows(t.rows) + ' rows' : '';
    // A plate behind the text, not a halo around the glyphs. An outline stroke
    // only covers the line where it runs under a letter; between the letters it
    // still shows, and a horizontal arrow reads as a strikethrough.
    const chars = Math.max(t.label.length, rows.length);
    const w = chars * CHARW + 8;
    svg.appendChild(el('rect', {
      x: t.x - w + 4, y: t.y - 10, width: w, height: rows ? 27 : 14, class: 'labelplate',
    }));
    svg.appendChild(el('text', { x: t.x, y: t.y, 'text-anchor': 'end', class: 'flabel' }, t.label));
    if (rows) {
      svg.appendChild(el('text', { x: t.x, y: t.y + 14, 'text-anchor': 'end', class: 'rowlabel' }, rows));
    }
  }
  svg.appendChild(arrowhead());

  for (const [symbol, p] of pos) {
    const n = p.node;
    const g = el('g', { class: 'flownode' });
    const failing = (n.tests || []).some(t => t.status === 'fail' || t.status === 'error');
    let cls = 'fn-model';
    if (n.kind === 'source') cls = 'fn-source';
    else if (n.kind === 'outside') cls = 'fn-outside';
    else if (n.changed) cls = 'fn-changed';
    if (failing) cls += ' fn-failing';

    g.appendChild(el('rect', { x: p.x, y: p.y, width: BOXW, height: p.h, rx: 4, class: 'fnbox ' + cls }));
    g.appendChild(el('text', { x: p.x + 12, y: p.y + 21, class: 'fntitle' }, truncName(n.name, 27)));

    let line = p.y + 39;
    const put = (text, klass) => {
      if (!text) return;
      g.appendChild(el('text', { x: p.x + 12, y: line, class: klass || 'blabel' }, trunc(text, 36)));
      line += LINE;
    };
    put(materialization(n), 'sp-svg-code');
    if (n.rows || n.bytes || n.nanos) {
      put([n.rows ? fmtRows(n.rows) + ' rows' : '', n.bytes ? fmtBytes(n.bytes) : '',
      n.nanos ? fmtNs(n.nanos) : ''].filter(Boolean).join(' · '), 'sp-svg-cost');
    }
    if (n.grain) put('one row per ' + n.grain, 'sp-svg-quote');
    if (n.unresolved) put('grain unreadable', 'sp-svg-risk');
    if (n.filter) put('where ' + n.filter, 'sp-svg-code');
    if ((n.aggregates || []).length) put(n.aggregates.join(', ') + ' — rows collapse', 'sp-svg-code');
    for (const t of n.tests || []) {
      const bad = t.status === 'fail' || t.status === 'error';
      put((bad ? '✗ ' : '✓ ') + t.name + (bad ? ' · ' + fmtRows(t.failures) + ' rows' : ''),
        bad ? 'sp-svg-risk' : 'sp-svg-ok');
    }
    if ((n.risks || []).length) put('! ' + n.risks.length + ' risk marker' + (n.risks.length > 1 ? 's' : ''), 'sp-svg-risk');

    g.appendChild(el('title', {}, flowTitle(n)));
    g.onclick = () => selectFlowNode(n);
    svg.appendChild(g);
  }
}

// A table written straight into the SQL has nothing upstream, so the
// longest-path layering puts it at 0, beside the sources, with its arrow
// crossing the whole picture and passing behind unrelated tables. It is not a
// source — it is an input to one model — so it is moved to sit beside that
// model's other inputs. The layer is presentation; the edge is unchanged.
function relayerOutsideTables(nodes, links) {
  const layerOf = new Map(nodes.map(n => [n.symbol, n.layer]));
  return nodes.map(n => {
    if (n.kind !== 'outside') return n;
    let deepest = 0;
    for (const l of links) {
      if (l.from === n.symbol) deepest = Math.max(deepest, layerOf.get(l.to) || 0);
    }
    return deepest > 1 ? { ...n, layer: deepest - 1 } : n;
  });
}

// orderColumns puts each table near the ones it connects to, so the arrows
// mostly do not cross. Sorting a column by name instead put the hardcoded
// refunds table at the top of its column while the model that reads it sat at
// the bottom of the next one, dragging a line across everything between.
//
// This is the barycentre sweep: order a column by the mean position of its
// neighbours in the column beside it, forwards then backwards, a few times.
// It is a heuristic and it does not promise zero crossings — but on a DAG this
// shape it removes the ones that make the picture hard to follow.
function orderColumns(layers, byLayer, links) {
  const index = new Map();
  const reindex = () => {
    for (const L of layers) byLayer.get(L).forEach((n, i) => index.set(n.symbol, i));
  };
  reindex();
  const neighbours = (n, back) => {
    const out = [];
    for (const l of links) {
      const other = back ? (l.to === n.symbol && l.from) : (l.from === n.symbol && l.to);
      if (other && index.has(other)) out.push(index.get(other));
    }
    return out;
  };
  for (let pass = 0; pass < 4; pass++) {
    const back = pass % 2 === 0;
    const order = back ? layers : [...layers].reverse();
    for (const L of order) {
      const col = byLayer.get(L);
      const key = new Map(col.map(n => {
        const ns = neighbours(n, back);
        // A table with nothing to anchor it keeps where it is, rather than
        // being sorted to one end for no reason.
        return [n.symbol, ns.length ? ns.reduce((a, b) => a + b, 0) / ns.length : index.get(n.symbol)];
      }));
      col.sort((a, b) => key.get(a.symbol) - key.get(b.symbol) || a.name.localeCompare(b.name));
      reindex();
    }
  }
}

// boxHeight sizes a table to what it has to say. Everything on a node is a fact
// from the run or the statement, so nothing is dropped to make the boxes tidy.
// LINE is the leading inside a table box, and boxHeight has to agree with it or
// the text runs out through the bottom edge.
const LINE = 14;

// The page is monospace throughout, so a string's width is its length. That is
// what makes a background plate sizeable without measuring text in the DOM.
const CHARW = 6.05;

function boxHeight(n) {
  let lines = 1; // materialization, always drawn
  if (n.rows || n.bytes || n.nanos) lines++;
  if (n.grain) lines++;
  if (n.unresolved) lines++;
  if (n.filter) lines++;
  if ((n.aggregates || []).length) lines++;
  lines += (n.tests || []).length;
  if ((n.risks || []).length) lines++;
  return 30 + lines * LINE + 10; // title band, the lines themselves, bottom padding
}

// truncName drops the front of a long qualified name, not the back. What
// identifies shop-prod-1234.shop_raw.refunds is "refunds"; cutting it to
// "shop-prod-1234.shop_raw.re…" throws away the only part you would recognise.
function truncName(s, n) {
  if (s.length <= n) return s;
  if (s.includes('.')) return '…' + s.slice(s.length - (n - 1));
  return trunc(s, n);
}

function materialization(n) {
  if (n.kind === 'source') return 'source';
  if (n.kind === 'outside') return 'outside the DAG';
  let m = n.materialized || 'model';
  if (n.unique_key) m += ' on ' + n.unique_key;
  if (n.status === 'not-run') m += ' · not rebuilt';
  return m;
}

// linkLabel is how the rows were matched. "left join on order_id" is the single
// most load-bearing string on the picture: it is the difference between rows
// being dropped, kept, or multiplied.
function linkLabel(l) {
  if (l.relation === 'from') return 'from';
  if (l.relation === 'ref') return 'declared, not in the statement';
  if (l.key) return l.relation + ' join on ' + l.key;
  if (l.on) return l.relation + ' join';
  return l.relation;
}

function flowTitle(n) {
  const out = [n.symbol];
  if (n.doc) out.push(n.doc);
  if (n.grain) out.push('grain: one row per ' + n.grain + ' (' + n.grain_from + ')');
  if (n.unresolved) out.push('grain unreadable — the SQL ' + n.unresolved);
  for (const r of n.risks || []) out.push('! ' + r);
  for (const t of n.tests || []) {
    out.push((t.status === 'fail' ? '✗ ' : '✓ ') + t.name +
      (t.failures ? ' — ' + fmtRows(t.failures) + ' rows failed' : ''));
  }
  return out.join('\n');
}

// Clicking a table copies its evidence, the same contract as clicking a frame:
// paste it into an agent and the question answers itself.
async function selectFlowNode(n) {
  const lines = [
    '## Model ' + n.name, '',
    materialization(n) + (n.rows ? ' · ' + fmtRows(n.rows) + ' rows' : '') +
    (n.bytes ? ' · ' + fmtBytes(n.bytes) + ' scanned' : '') + (n.nanos ? ' · ' + fmtNs(n.nanos) : ''),
  ];
  if (n.doc) lines.push('', n.doc);
  if (n.grain) lines.push('', 'Grain: one row per ' + n.grain + ' (' + n.grain_from + ')');
  if (n.unresolved) lines.push('', 'Grain could not be read: the SQL ' + n.unresolved);
  const ins = (DATA.flow.links || []).filter(l => l.to === n.symbol);
  if (ins.length) {
    lines.push('', '## Reads');
    for (const l of ins) {
      lines.push('- ' + l.from_name + ' — ' + linkLabel(l) + (l.note ? ': ' + l.note : '') +
        (l.in_dag ? '' : ' (written into the SQL, invisible to dbt)'));
    }
  }
  const outs = (DATA.flow.links || []).filter(l => l.from === n.symbol);
  if (outs.length) {
    lines.push('', '## Read by');
    for (const l of outs) lines.push('- ' + l.to_name + ' — ' + linkLabel(l));
  }
  if ((n.tests || []).length) {
    lines.push('', '## Tests');
    for (const t of n.tests) {
      lines.push('- ' + t.name + ' — ' + t.status + (t.failures ? ' (' + fmtRows(t.failures) + ' rows failed)' : ''));
    }
  }
  if ((n.risks || []).length) {
    lines.push('', '## Risk markers');
    for (const r of n.risks) lines.push('- ' + r);
  }
  await select(n.symbol, null, { copy: false });
  const pc = await (await fetch('/api/symbol/' + encodeURIComponent(n.symbol))).json();
  await copy(lines.join('\n') + '\n\n---\n\n' + (pc.markdown || ''),
    n.unresolved ? 'model copied — grain unreadable, paste it and ask' : 'model evidence copied');
}

function fmtRows(n) {
  if (!n) return '0';
  return String(n).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

function fmtBytes(n) {
  if (!n) return '';
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
  let i = 0, v = n;
  while (v >= 1024 && i < units.length - 1) { v /= 1024; i++; }
  return (i === 0 ? v : v.toFixed(1)) + ' ' + units[i];
}

function arrowhead() {
  const defs = el('defs', {});
  const m = el('marker', {
    id: 'arrow', viewBox: '0 0 8 8', refX: 7, refY: 4,
    markerWidth: 7, markerHeight: 7, orient: 'auto-start-reverse',
  });
  m.appendChild(el('path', { d: 'M0,0 L8,4 L0,8 z', class: 'arrowhead' }));
  defs.appendChild(m);
  return defs;
}

// flowSummary is the header line: what the build cost and whether to trust it.
function flowSummary(flow) {
  const bits = [];
  if (flow.elapsed_ns) bits.push(fmtNs(flow.elapsed_ns) + ' elapsed');
  if (flow.bytes_scanned) bits.push(fmtBytes(flow.bytes_scanned) + ' scanned');
  if (flow.rows_written) bits.push(fmtRows(flow.rows_written) + ' rows written');
  let text = 'Build order, left to right: every table sits to the right of everything it reads. ' + bits.join(' · ') + '.';
  if (flow.failing) {
    text += ' ' + flow.failing + ' test' + (flow.failing > 1 ? 's are' : ' is') +
      ' failing, so at least one of these tables is not what it claims to be.';
  }
  return text;
}

File: internal/lang/dbt/flow.go
package dbt

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/kelalaike/plum/internal/bundle"
)

// A warehouse build is not a call stack.
//
// The landscape draws code: a path that descends into a call and comes back
// with a value, and closure is the shape you read it by. SQL has no returns.
// fct_orders does not call stg_orders and get control back — the warehouse
// builds stg_orders, builds stg_payments, then builds fct_orders by reading
// both. Drawing that as a path invents resumes that never happened and buries
// the only structural fact that matters, which is that two tables met.
//
// So dbt gets its own picture: layered by dependency depth, data moving one
// way, every arrow carrying how the rows were matched and what came through.

type Flow struct {
	SessionID    string     `json:"session_id"`
	Nodes        []FlowNode `json:"nodes"`
	Links        []FlowLink `json:"links"`
	ElapsedNanos int64      `json:"elapsed_ns"`
	BytesScanned int64      `json:"bytes_scanned"`
	RowsWritten  int64      `json:"rows_written"`
	Failing      int        `json:"failing"`
	Findings     []string   `json:"findings,omitempty"`
}

type FlowNode struct {
	Symbol bundle.SymbolID `json:"symbol"`
	Name   string          `json:"name"`
	File   string          `json:"file"`
	// Layer is the longest path from a source, so everything a node reads is
	// strictly to its left. That is the build order, and it is why the picture
	// can be read without following arrows backwards.
	Layer        int    `json:"layer"`
	Kind         string `json:"kind"` // source | model
	Materialized string `json:"materialized,omitempty"`
	UniqueKey    string `json:"unique_key,omitempty"`
	Rows         int64  `json:"rows,omitempty"`
	Bytes        int64  `json:"bytes,omitempty"`
	Nanos        int64  `json:"nanos,omitempty"`
	Status       string `json:"status"` // success | fail | error | skipped | not-run
	Changed      bool   `json:"changed"`
	Doc          string `json:"doc,omitempty"`

	// Grain is what one row of this table means. GrainFrom says who says so:
	// "declared" is the author's prose, "inferred" is what the SQL does. The
	// distinction is the whole point — one is a promise, the other is evidence.
	Grain      string `json:"grain,omitempty"`
	GrainFrom  string `json:"grain_from,omitempty"`
	Unresolved string `json:"unresolved,omitempty"`

	Filter     string     `json:"filter,omitempty"`
	Aggregates []string   `json:"aggregates,omitempty"`
	Tests      []FlowTest `json:"tests,omitempty"`
	Risks      []string   `json:"risks,omitempty"`
}

type FlowTest struct {
	Name     string `json:"name"`
	Status   string `json:"status"`
	Failures int    `json:"failures,omitempty"`
	Nanos    int64  `json:"nanos,omitempty"`
	Bytes    int64  `json:"bytes,omitempty"`
}

// FlowLink is one table meeting another. How they were matched is the fact a
// call arrow cannot carry: an inner join drops rows that do not match, a left
// join keeps them, and a join on a non-unique key multiplies them.
type FlowLink struct {
	From     bundle.SymbolID `json:"from"`
	To       bundle.SymbolID `json:"to"`
	FromName string          `json:"from_name"`
	ToName   string          `json:"to_name"`
	// Relation is how the target reads this input:
	//   from   the driving table of the select
	//   inner | left | right | full | cross   the join that brought it in
	//   ref    declared in the DAG but not found in the statement
	Relation string `json:"relation"`
	Key      string `json:"key,omitempty"` // the matched column, when it is a plain equality
	On       string `json:"on,omitempty"`  // the condition as written
	Rows     int64  `json:"rows,omitempty"`
	// InDAG is false for a table written straight into the SQL. dbt does not
	// know about it, will not build it first, and will not show it in lineage —
	// so it is drawn, in the one place that can see it.
	InDAG bool `json:"in_dag"`
	// Note is what this edge does to the row count, in words.
	Note string `json:"note,omitempty"`
}

// BuildFlow reads a dbt run and the SQL behind it. root is the project
// directory, used to read the statements themselves — the run says what a query
// cost, only the statement says what it did.
func BuildFlow(m *Manifest, r *RunResults, root string, changed map[bundle.SymbolID]bool, risks map[bundle.SymbolID][]string) *Flow {
	byID := map[string]result{}
	for _, res := range r.Results {
		byID[res.UniqueID] = res
	}
	f := &Flow{ElapsedNanos: int64(r.ElapsedTime * float64(time.Second))}

	shapes := map[string]Shape{}
	index := map[string]bundle.SymbolID{} // model name → symbol

	// Nodes: every model and source the manifest knows, whether or not this run
	// touched it. A model that did not rebuild is still upstream of one that did.
	var ids []string
	for id, n := range m.Nodes {
		if n.ResourceType == "test" {
			continue
		}
		ids = append(ids, id)
	}
	sort.Strings(ids)

	for _, id := range ids {
		n := m.Nodes[id]
		sym := m.symbolID(id)
		index[n.Name] = sym
		node := FlowNode{
			Symbol: sym, Name: displayName(id, n), File: n.OriginalFilePath,
			Kind: n.ResourceType, Materialized: n.Config.Materialized,
			UniqueKey: n.Config.UniqueKey, Doc: n.Description,
			Changed: changed[sym], Risks: risks[sym],
			Status: "not-run",
		}
		if res, ran := byID[id]; ran {
			node.Status = res.Status
			node.Rows = res.Adapter.RowsAffected
			node.Bytes = res.Adapter.BytesProcessed
			node.Nanos = int64(res.ExecutionTime * float64(time.Second))
			f.BytesScanned += node.Bytes
			f.RowsWritten += node.Rows
		}
		if n.ResourceType == "model" && n.OriginalFilePath != "" {
			if src, err := os.ReadFile(filepath.Join(root, n.OriginalFilePath)); err == nil {
				sh := ReadShape(string(src), n.Description+"\n"+leadingComment(string(src)))
				shapes[id] = sh
				node.Filter = sh.Where
				node.Aggregates = sh.Aggregates
				node.Unresolved = sh.Unresolved
				switch {
				case sh.Inferred != "":
					node.Grain, node.GrainFrom = sh.Inferred, "inferred"
				case sh.Declared != "":
					node.Grain, node.GrainFrom = sh.Declared, "declared"
				}
				if d := sh.GrainDivergence(); d != "" {
					f.Findings = append(f.Findings, n.Name+": "+d)
				}
				if sh.Unresolved != "" {
					f.Findings = append(f.Findings, n.Name+": "+sh.Unresolved)
				}
			}
		}
		f.Nodes = append(f.Nodes, node)
	}

	// Tests hang off the model they check. A test is a statement about a table,
	// not a step in the build, and drawing it as a frame was always a fiction.
	nodeAt := map[bundle.SymbolID]int{}
	for i, n := range f.Nodes {
		nodeAt[n.Symbol] = i
	}
	for id, n := range m.Nodes {
		if n.ResourceType != "test" {
			continue
		}
		res, ran := byID[id]
		if !ran {
			continue
		}
		t := FlowTest{Name: n.Name, Status: res.Status, Nanos: int64(res.ExecutionTime * float64(time.Second)), Bytes: res.Adapter.BytesProcessed}
		if res.Failures != nil {
			t.Failures = *res.Failures
		}
		if t.Status == "fail" || t.Status == "error" {
			f.Failing++
		}
		// A test is a query. It scans billed bytes like any other, so leaving it
		// out of the total would understate what the run cost.
		f.BytesScanned += t.Bytes
		for _, dep := range n.DependsOn.Nodes {
			if i, ok := nodeAt[m.symbolID(dep)]; ok {
				f.Nodes[i].Tests = append(f.Nodes[i].Tests, t)
			}
		}
	}
	for i := range f.Nodes {
		sort.Slice(f.Nodes[i].Tests, func(a, b int) bool { return f.Nodes[i].Tests[a].Name < f.Nodes[i].Tests[b].Name })
	}

	// Links: the DAG dbt declares, annotated with what the statement does. A
	// dependency the statement does not mention is still drawn — it is real —
	// but it is marked as declared rather than read.
	for _, id := range ids {
		n := m.Nodes[id]
		to := m.symbolID(id)
		sh := shapes[id]
		matched := map[string]bool{}
		for _, dep := range n.DependsOn.Nodes {
			depNode, ok := m.Nodes[dep]
			if !ok {
				continue
			}
			from := m.symbolID(dep)
			link := FlowLink{
				From: from, To: to, FromName: displayName(dep, depNode), ToName: n.Name,
				Relation: "ref", InDAG: true, Rows: byID[dep].Adapter.RowsAffected,
			}
			switch {
			case sh.From != "" && sameTarget(sh.From, depNode.Name):
				link.Relation = "from"
				link.Note = "the driving table"
				matched[sh.From] = true
			default:
				for _, j := range sh.Joins {
					if sameTarget(j.Target, depNode.Name) {
						link.Relation, link.Key, link.On = j.Type, j.Key, j.On
						link.Note = joinNote(j)
						matched[j.Target] = true
						break
					}
				}
			}
			f.Links = append(f.Links, link)
		}
		// A table written straight into the SQL is read by this model and is
		// invisible to dbt. It is the one edge the manifest cannot tell you
		// about, so it is exactly the one worth drawing.
		for _, j := range sh.Joins {
			if matched[j.Target] || index[j.Target] != "" || !strings.Contains(j.Target, ".") {
				continue
			}
			outside := bundle.SymbolID("::" + j.Target)
			if _, seen := nodeAt[outside]; !seen {
				nodeAt[outside] = len(f.Nodes)
				f.Nodes = append(f.Nodes, FlowNode{
					Symbol: outside, Name: j.Target, Kind: "outside", Status: "not-run",
					Risks: []string{"written into the SQL, so dbt will not build it first or track it in lineage"},
				})
			}
			f.Links = append(f.Links, FlowLink{
				From: outside, To: to, FromName: j.Target, ToName: n.Name,
				Relation: j.Type, Key: j.Key, On: j.On, Note: joinNote(j), InDAG: false,
			})
			f.Findings = append(f.Findings, n.Name+": reads "+j.Target+" directly, outside the DAG")
		}
	}

	// Layer is the longest path from a node with nothing upstream, so an arrow
	// never points backwards and the build order reads left to right.
	inbound := map[bundle.SymbolID][]bundle.SymbolID{}
	for _, l := range f.Links {
		inbound[l.To] = append(inbound[l.To], l.From)
	}
	var depth func(bundle.SymbolID, map[bundle.SymbolID]bool) int
	depth = func(s bundle.SymbolID, seen map[bundle.SymbolID]bool) int {
		if seen[s] {
			return 0 // a cycle dbt would have rejected; do not spin on it
		}
		seen[s] = true
		defer delete(seen, s)
		best := 0
		for _, up := range inbound[s] {
			if d := depth(up, seen) + 1; d > best {
				best = d
			}
		}
		return best
	}
	for i := range f.Nodes {
		f.Nodes[i].Layer = depth(f.Nodes[i].Symbol, map[bundle.SymbolID]bool{})
	}
	sort.SliceStable(f.Nodes, func(a, b int) bool {
		if f.Nodes[a].Layer != f.Nodes[b].Layer {
			return f.Nodes[a].Layer < f.Nodes[b].Layer
		}
		return f.Nodes[a].Name < f.Nodes[b].Name
	})
	sort.Strings(f.Findings)
	return f
}

// displayName is what to call a node on the picture. A source is written into
// the SQL as schema.table, and naming it by the table alone drops the half that
// says which system it came out of.
func displayName(id string, n node) string {
	if n.ResourceType != "source" {
		return n.Name
	}
	parts := strings.Split(id, ".")
	if len(parts) >= 2 {
		return strings.Join(parts[len(parts)-2:], ".")
	}
	return n.Name
}

// joinNote says what the join does to the row count, which is the question
// somebody reading a mart actually has.
func joinNote(j Join) string {
	switch j.Type {
	case "inner":
		return "rows without a match on both sides are dropped"
	case "left":
		return "unmatched rows are kept, and rows multiply if the key repeats"
	case "right":
		return "rows on this side are kept, the other side's unmatched rows are dropped"
	case "full":
		return "both sides are kept, unmatched columns come back null"
	case "cross":
		return "every row meets every row"
	}
	return ""
}

// sameTarget compares a name written in SQL with a name from the manifest. A
// source is written shop_raw.orders and named orders.
func sameTarget(sqlName, nodeName string) bool {
	if strings.EqualFold(sqlName, nodeName) {
		return true
	}
	return strings.EqualFold(lastSegmentSQL(sqlName), nodeName)
}

func (f *Flow) Save(path string) error {
	data, err := json.MarshalIndent(f, "", "  ")
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

func LoadFlow(path string) (*Flow, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var f Flow
	if err := json.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("parsing flow: %w", err)
	}
	return &f, nil
}
